src/boot/efi/cpio.c

```diff
@@ -0,0 +1,465 @@
+        a += 6;
+
+        a = write_cpio_word(a, (*inode_counter)++);                         /* inode */
+        a = write_cpio_word(a, access_mode | 0040000 /* = S_IFDIR */);      /* mode */
+        a = write_cpio_word(a, 0);                                          /* uid */
+        a = write_cpio_word(a, 0);                                          /* gid */
+        a = write_cpio_word(a, 1);                                          /* nlink */
+        a = write_cpio_word(a, 0);                                          /* mtime */
+        a = write_cpio_word(a, 0);                                          /* size */
+        a = write_cpio_word(a, 0);                                          /* major(dev) */
+        a = write_cpio_word(a, 0);                                          /* minor(dev) */
+        a = write_cpio_word(a, 0);                                          /* major(rdev) */
+        a = write_cpio_word(a, 0);                                          /* minor(rdev) */
+        a = write_cpio_word(a, path_size + 1);                              /* fname size */
+        a = write_cpio_word(a, 0);                                          /* "crc" */
+
+        CopyMem(a, path, path_size + 1);
+        a += path_size + 1;
+
+        /* Pad to next multiple of 4 */
+        a = pad4(a, *cpio_buffer);
+
+        assert(a == (CHAR8*) *cpio_buffer + *cpio_buffer_size + l);
+
+        *cpio_buffer_size += l;
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS pack_cpio_prefix(
+                const CHAR8 *path,
+                UINT32 dir_mode,
+                UINT32 *inode_counter,
+                VOID **cpio_buffer,
+                UINTN *cpio_buffer_size) {
+
+        EFI_STATUS err;
+        const CHAR8 *p = path;
+
+        /* Serializes directory inodes of all prefix paths of the specified path in cpio format. Note that
+         * (similar to mkdir -p behaviour) all leading paths are created with 0555 access mode, only the
+         * final dir is created with the specified directory access mode. */
+
+        for (;;) {
+                const CHAR8 *e;
+
+                e = strchra(p, '/');
+                if (!e)
+                        break;
+
+                if (e > p) {
+                        _cleanup_freepool_ CHAR8 *t = NULL;
+
+                        t = strndup8(path, e - path);
+                        if (!t)
+                                return EFI_OUT_OF_RESOURCES;
+
+                        err = pack_cpio_dir(t, 0555, inode_counter, cpio_buffer, cpio_buffer_size);
+                        if (EFI_ERROR(err))
+                                return err;
+                }
+
+                p = e + 1;
+        }
+
+        return pack_cpio_dir(path, dir_mode, inode_counter, cpio_buffer, cpio_buffer_size);
+}
+
+
+static EFI_STATUS pack_cpio_trailer(
+                VOID **cpio_buffer,
+                UINTN *cpio_buffer_size) {
+
+        static const char trailer[] =
```

Comment:
Shouldn't this just be normal string constant instead, do we need an array?

Reply:
a static const char array is exactly the very same as a literal string in the resulting ELF binary. 

I mean I could use a `#define` for the string here, to give it a name (and we want to give it a name, since we need to refer to it at various times, e.g. we use sizeof() on it multiple times). But `#define` would define a name for it for the rest of the file. By doing a static array OTOH the name is only valid within this code block, and not beyond.